### DISCUSSION TITLE
[WIP] changes for next release 3.11.0

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#3243] [Announcements]
This thread collects all changes made since the 3.10.5 (#3238) release.

#### New features

- Allow to **use `std::string_view` as object keys** in `at`, `operator[]`, `value`, `erase`, `find`, `count`, and `contains` to avoid unnecessary allocations.
  - #3423
  - #1529
  - #3558
  - #3564
- Add support to read and write the UBJSON-derived **[Binary JData (BJData)](https://github.com/NeuroJSON/bjdata/blob/Draft_2/Binary_JData_Specification.md) format**.
  - #3336
  - #3461
  - #3463
  - #3475
  - #3479
  - #3493
  - #3491
  - #3492
  - #3490
  - #3500
  - #3502
  - #3503
  - #3505
  - #3513
  - #3514
  - #3519
  - #3523
  - #3541
  - #3543

#### Bug fixes

- fix creation of `ordered_json` object from initializer list
  - #3342
  - #3343
  - #3370
- fix compilation problems in C++20/GCC 12
  - #3138
  - #3379
- make iterator satisfy `std::incrementable`
  - #3331
  - #3332
- exclude std::any from implicit conversion
  - #3428
  - #3437
- fix constraints on from_json() for strings
  - #3171
  - #3267
  - #3312
  - #3384
  - #3427
  - #3312
  - #3620
- Iterators cannot be used with range-v3
  - #3130
  - #3446
- Tests failing when compiling with c++20
  - #3207
  - #3446
  - #3472
- Comparison of NaN differs between json and float
  - #3409
  - #3446
- Re-add `value_type` detection to distinguish string types
  - #3204
  - #3333
  - #3604
  - #3602
  - #3629
- Add `operator<<(json_pointer)` to fix regression after 3.10.0 release
  - #3600
  - #3601
- Fix: Applying JSON patch creates parent object
  - #3134
  - #3199
  - #3628
- Fixed latest build error in MSVC platform
  - #3630

#### Improvements

- Allow default values for NLOHMANN_DEFINE_TYPE_INTRUSIVE and NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE
  - #2819
  - #3143
- Avoid clash with Arduino defines
  - #3338
- improve `meta()` for MSVC
  - #3417 
- check if different version of library is already included
  - #3418
- re-template json_pointer on string type
  - #3415
- allow to create booleans from `std::vector<bool>::reference`
  - #3533
  - #3534
- allow to use array types where iterators are pointers
  - #3544
- fix a `-Wpragmas` warning in GCC <11
  - #3550
- optimize performance of output vector adapter
  - #3569
- add `patch_inplace` function to apply patches without copying
  - #3581
  - #3596
- Use `swap` by ADL to allow swapping with non-`std` containers; also improve error messages
  - #3609
- Add versioned, ABI-tagged inline namespace
  - #1539
  - #3360
  - #3588 
  - #3590
- Move UDLs out of global namespace
  - #1682
  - #3605
- Fix ICPC warning 1098:
  - #3632
  - #3634

#### Further Changes

- fix _MSC_VER version to check for std::filesystem
  - #3240
- remove stringstream usage
  - #3239 
  - #3244
- add more examples to documentation
  - #3464
- improve documentation on supported types of `parse` and `accept`
  - #3245 
  - #3246
- add documentation on how to parse JSON Lines input
  - #3247
- fix typos
  - #3249
  - #3265
  - #3426
  - #3439
  - #3481
  - #3499
- add maintainer targets to create source archive
  - #3289
  - #3255
- add support for Visual Studio 2022
  - #3295
- document parsing to `ordered_json`
  - #3325
  - #3326 
- fix CITATION.cff
  - #3320 
- adjust GitHub Actions CI wrt. `windows-latest` image (by @falbrechtskirchinger)
  - #3368
- use `FetchContent_MakeAvailable` in examples
  - #3345
  - #3351
- document requirement of using the same definition of `JSON_DIAGNOSTICS` in all linked objects
  - #3360
  - #3378
- improve unit tests
  - #3380 
  - #3393 
  - #3365 
  - #3405 
  - #3377
  - #3421
  - #3422
- overwork bug template
  - #3348
- remove deprecated windows-2016 image
  - #3416
- speed up AppVeyor CI
  - #3422 
- update CI image
  - #3420
- add script to test local version in Compiler Explorer
  - #3456
- use [The Pitchfork Layout (PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs)
  - #3462
- add documentation of macros and runtime assertions
  - #3444 
  - #3431
- update Cpplint
  - #3454
- disable regression test on GCC <8.4
  - #3451
- add build step for ICPC to CI
  - #3465
- documented fuzz testing
  - #3477
  - #3478
- add more Xcode versions to the CI
  - #3485
- fix documentation of JSON Pointer
  - #3511
  - #3520 
- add tests for 32 bit
  - #3524
  - #3529
  - #3530
  - #3532
- change the default value CMake parameter `JSON_MultipleHeader` to `ON` to always use the multi-header version unless specified otherwise
  - #3532
- add option to disable default enum conversions
  - #3536
- fix "JSON_MultipleHeaders" option spelling in CI
  - #3555
- more intermediate folders from Visual Studio Code are added to `.gitignore`
  - #3572
- increase timeout for Unicode tests
  - #3579
  - #3580
  - #3614
- add badge for https://repology.org/project/nlohmann-json/versions
  - #3586
- make certain usage patterns more prominent in the README
  - #3557
- use [REUSE licensing framework](http://reuse.software)
  - #3546 
  - #3633
  - #3635
- add error message if test suite cannot be found
  - #3584
  - #3585
- document precondition for parsing from `FILE *` and add assertion
  - #3593 
- improve documentation
  - #3592
- add unit test to make sure `iterator_input_adapter` advances iterators correctly
  - #3548
- Add documentation for comparing `json` and `ordered_json`
  - #3443
  - #3599
- About CI runs on newer commits on the same branch
  - #3610
- Fix MinGW CI failures
  - #3618
- Install `.pc` and `.cmake` files to `share` directory
  - #3619
- Adjust JSON Pointer examples and add CI step for the examples
  - #3622
- Disable exceptions on ICPC
  - #3621
- Add a unit test including `<windows.h>`
  - #3631
- Overwork documentation and add more examples
  - #3553

**Comment by engine5d** [comment]
hi, when will be the release 3.11.0 published?

**Reply by nlohmann** [reply]
I would like to see (at least) https://github.com/nlohmann/json/pull/3446, https://github.com/nlohmann/json/pull/3423, and https://github.com/nlohmann/json/pull/3336 merged before making the release.

**Reply by gregmarr** [reply]
I think #3442 would be good too.

**Reply by falbrechtskirchinger** [reply]
@nlohmann How close are we to a release? If it's imminent I'll have to adjust my plans.

**Reply by nlohmann** [reply]
I am giving a talk on Qt DevCon (https://www.qtdevcon.com/25k-stars-later/) on June 14 and would like to release before (maybe June 12 or 13). So take your time. What do you have in mind?

**Reply by falbrechtskirchinger** [reply]
I'd like to scan through my WIP branches for small things that can/should go into 3.11.0. For example, I'd like to at least include an assertion to deal with #3425/#2649. I'm also aware of a compile issue with `-Werror=strict-overflow` on GCC 12 newer than the one in the `json-ci` image.
Slightly bigger things like API parity ("missing" `json_pointer` overloads) would also be nice.

**Reply by nlohmann** [reply]
The 3.11.0 release is already way too late. As always, I would like to release more often, but something always comes up...

**Reply by falbrechtskirchinger** [reply]
I completely understand. I'll make an effort to get the small things (couple of lines each) in over the weekend.

**Reply by falbrechtskirchinger** [reply]
@nlohmann Did you want to mention the "JSON_MultipleHeader=ON" change (part of #3532) specifically?

**Reply by nlohmann** [reply]
Thanks, done.

**Comment by nlohmann** [comment]
@falbrechtskirchinger @gregmarr  What would you like to see in the 3.11.0? I am thinking of:

- #3553
- #3590

Apart of that, I think we should have a kind of feature-freeze and focus on documentation (and release notes) for a while and get the 3.11.0 finally through the door.

I am traveling next week and should have some time in the first week of August.

What do you think?

**Reply by falbrechtskirchinger** [reply]
Sounds good. While there are other PRs relatively close to being ready for merge, they are probably not as critical. Also, 3.11.0 is seriously overdue.

There's one more item I'd like you to take care of. I believe you create the release version tags via your release scripts? Right now, you are tagging the merge commit, which doesn't sit on the `develop` branch. This means the tags are less useful than they could be, e.g., they don't show up in `git describe --tags` which currently prints `v3.10.1-117-gd4daaa897`. Unless there's a good reason I'm missing, could you please tag the commit preceding the merge commit? (Ideally, you'd look into changing the tags for 3.10.2-3.10.5 retroactively, as well. Don't know if that'd create issues with GitHub.)

**Reply by nlohmann** [reply]
The release script uses git flow and creates a release branch.

See: https://github.com/nlohmann/json_release_scripts/blob/main/scripts/close_branches/close_branches.sh#L12

That command creates the tag. If that does not make sense, then I'm confused.

**Reply by falbrechtskirchinger** [reply]
Creating a release branch is fine, but the tag should be associated with the commit residing on the `develop` branch (and the release branch), not the merge commit **only** residing on the release branch. That way the release history is visible on the `develop` branch.

I'll take a look at `git flow` later and see if there's anything to be done.

**Reply by falbrechtskirchinger** [reply]
I'm not alone. This has been discussed in https://github.com/nvie/gitflow/issues/49 and was addressed *somehow*. Trying to figure out what flag you'd need to set.

**Reply by falbrechtskirchinger** [reply]
Could you tell me what version you're using?

```sh
$ git flow version
1.12.3 (AVH Edition)
```

**Reply by nlohmann** [reply]
I'm using Homebrew's version

```
➜  git flow version
0.4.1
➜  brew info git-flow
git-flow: stable 0.4.1 (bottled), HEAD
```

**Reply by falbrechtskirchinger** [reply]
That explains things. That version was tagged `Feb 14, 2011`. And that probably explains why Linux distributions are shipping `AVH Edition` which seems to have continued development.

It's also available on Homebew:
https://formulae.brew.sh/formula/git-flow-avh

**Reply by nlohmann** [reply]
I updated on my side and adjusted the release script.

**Reply by nlohmann** [reply]
I will be traveling in the next days. Until then, I will try to get the release script up running so that this is not a blocked once everything is merged and the release notes (...) are done.

**Reply by falbrechtskirchinger** [reply]
Great! We should get 3.11.0 out the door next week. :-)
(Reminder that Wednesday is a MacOS 10.5 brownout day, IIRC.)

**Reply by nlohmann** [reply]
I think we have it now. The following PRs are in the queue and will be merged once the CI succeeds:

- #3629 
- #3631
- #3633
- #3634

I'll try to create a pre-release tomorrow or on Monday. I'll refine the list of issues/PRs above to release notes to get this through the door until Wednesday.

**Reply by falbrechtskirchinger** [reply]
What about these?
* #3553
* #3605

**Reply by nlohmann** [reply]
I forgot about #3553. I am not sure about #3605. Do we really need an option? Can’t we just handle this like in #3590?

**Reply by falbrechtskirchinger** [reply]
We can't do so unconditionally without breaking code. And I suspect you'd want me to change the default to `JSON_USE_GLOBAL_UDLS=1` – the commits for that change are ready, but I wanted to see the impact of `JSON_USE_GLOBAL_UDLS=0`.

Also, I made some mistakes yesterday and mixed up the UDL documentation between those PRs. I'm now waiting for a decision before I fix that.

**Reply by nlohmann** [reply]
I’ll check later. In any case, the change should be non-breaking and least surprising :)

**Reply by falbrechtskirchinger** [reply]
Alright, I've decided to undo the documentation changes pertaining to UDLs in #3553. I'll also change the defaults in #3605 to be non-breaking and remove the draft status when finished.

After #3553 has been merged and I've rebased #3605, I'll re-apply the UDL-related documentation changes. You can decide the fate of #3605 then. :-)

**Reply by nlohmann** [reply]
> I'm not alone. This has been discussed in [nvie/gitflow#49](https://github.com/nvie/gitflow/issues/49) and was addressed _somehow_. Trying to figure out what flag you'd need to set.

Two questions:

- Is the tag's commit now as you expected?
- I realized that the repo's `master` branch is gone. Is that expected with the newer git-flow?

**Reply by nlohmann** [reply]
Even worse, my release script fails now. :-/

**Reply by falbrechtskirchinger** [reply]
Since I don't use `git flow` myself, I've no idea.

**Reply by falbrechtskirchinger** [reply]
Oh, and yes. :-) `git describe --tags` is producing the output I expect.

**Reply by nlohmann** [reply]
> Since I don't use `git flow` myself, I've no idea.

After the 3.11.1 release, there is still a `master` branch, so that seems to work now.

**Comment by falbrechtskirchinger** [comment]
#3199 is another nice-to-have bug fix for 3.11.0. I might update that PR over the weekend since the original author hasn't responded.

**Comment by nlohmann** [comment]
All planned PRs are merged. Here are my proposed release notes - feedback more than welcome (cc @falbrechtskirchinger @gregmarr). Going to bed now and hoping to get this out tomorrow. Note in GitHub's release page, issues/PRs are just linked with a number and not with icon/title like here.

* * *

### Summary

Version 3.11.0 one of the biggest release (in terms of closed issues and merged pull requests) ever. This version brings:

- **String view** support for all functions working on object keys (e.g., [`at`](https://json.nlohmann.me/api/basic_json/at/) or [`operator[]`](https://json.nlohmann.me/api/basic_json/operator%5B%5D/)). This allows to avoid unnecessary allocations.
- **BJData** as fifth supported binary format beside BSON, CBOR, MessagePack, and UBJSON.
- **Better C++20 support** to make the library integrate more smoothly to newer code bases.
- Better support for avoiding problems when **multiple versions** of the library are used in the same project.
- Even better and more extensive **documentation** and **examples**.
- **More tests** for edge cases like 32-bit support, C++20 features, using the library with ICPC, or after including `<windows.h>`.

All changes are backward-compatible.

:moneybag: Note you can **support this project** via [GitHub sponsors](https://github.com/sponsors/nlohmann) or [PayPal](https://paypal.me/nlohmann).

### :sparkles: New Features

- Allow to **use `std::string_view` as object keys** in [`at`](https://json.nlohmann.me/api/basic_json/at/), [`operator[]`](https://json.nlohmann.me/api/basic_json/operator%5B%5D/), [`value`](https://json.nlohmann.me/api/basic_json/value/), [`erase`](https://json.nlohmann.me/api/basic_json/erase/), [`find`](https://json.nlohmann.me/api/basic_json/find/), [`contains`](https://json.nlohmann.me/api/basic_json/contains/), and [`count`](https://json.nlohmann.me/api/basic_json/count/) to avoid unnecessary allocations. #3423, #1529, #3558, #3564
- Add support to read and write the UBJSON-derived **[Binary JData (BJData)](https://github.com/NeuroJSON/bjdata/blob/Draft_2/Binary_JData_Specification.md) format** (see [documentation](https://json.nlohmann.me/features/binary_formats/bjdata/)). #3336, #3461, #3463, #3475, #3479, #3493, #3491, #3492, #3490, #3500, #3502, #3503, #3505, #3513, #3514, #3519, #3523, #3541, #3543

### :bug: Bug fixes

- Allow creation of [`ordered_json`](https://json.nlohmann.me/api/ordered_json/) objects from initializer lists. #3342, #3343, #3370
- Fix failing tests when compiling with C++20 and add support for [`operator<=>`](https://json.nlohmann.me/api/basic_json/operator_spaceship/). #3207, #3446, #3472
- Fix comparison-related compilation problems in C++20 on GCC 12. #3138, #3379
- Make iterator satisfy `std::incrementable`. #3331, #3332
- Exclude `std::any` from implicit conversion. #3428, #3437
- Fix constraints on `from_json()` for strings. #3171, #3267, #3312, #3384, #3427, #3312, #3620
- Make iterators usable with range-v3. #3130, #3446
- Fix comparison of NaN values to `json` to behave like `float`. #3409, #3446
- Add [`operator<<(json_pointer)`](https://json.nlohmann.me/api/operator_ltlt/) to fix regression after 3.10.0 release. #3600, #3601
- Fix JSON Patch to not create missing parent objects. #3134, #3199, #3628
- Re-add `value_type` detection to distinguish string types (was broken after 3.10.5 release). #3204, #3333, #3604, #3602, #3629
- Fix latest build error in MSVC platform (was broken after 3.10.5 release). #3630

### :zap: Improvements

- Allow default values for [`NLOHMANN_DEFINE_TYPE_INTRUSIVE`](https://json.nlohmann.me/api/macros/nlohmann_define_type_intrusive/) and [`NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE`](https://json.nlohmann.me/api/macros/nlohmann_define_type_non_intrusive/). #2819, #3143
- Avoid clash with Arduino defines. #3338
- Improve [`meta`](https://json.nlohmann.me/api/basic_json/meta/) output for MSVC. #3417 
- Check and warn if different version of library is already included (see [`JSON_SKIP_LIBRARY_VERSION_CHECK`](https://json.nlohmann.me/api/macros/json_skip_library_version_check/)). #3418
- Re-template [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) on string type. #3415
- Allow to create booleans from `std::vector<bool>::reference`. #3533, #3534
- Allow to use array types where iterators are pointers. #3544
- Improve performance of writing binary formats to byte vectors. #3569
- Add [`patch_inplace`](https://json.nlohmann.me/api/basic_json/patch_inplace/) function to apply patches without copying. #3581, #3596
- Use [`swap`](https://json.nlohmann.me/api/basic_json/swap/) by ADL to allow swapping with non-`std` containers and improve error messages. #3609
- Add versioned, ABI-tagged inline namespace (see [`NLOHMANN_JSON_NAMESPACE`](https://json.nlohmann.me/api/macros/nlohmann_json_namespace/)) to avoid ABI errors when linking different versions of the library. #1539, #3360, #3588, #3590
- Move UDLs out of global namespace (see [`JSON_USE_GLOBAL_UDLS`](https://json.nlohmann.me/api/macros/json_use_global_udls/)) to avoid ABI errors when linking different versions of the library. #1682, #3605

#### Warnings

- Fix ICPC warning 1098. #3632, #3634
- Fix a `-Wpragmas` warning in GCC <11. #3550

### :hammer: Further Changes

- Fix `_MSC_VER` version to check for `std::filesystem`. #3240
- Remove `<sstream>` dependency. #3239, #3244
- Overwork bug template to systematically request information. #3348
- Add script to [test local version](https://github.com/nlohmann/json/tree/develop/tools/serve_header) in Compiler Explorer. #3456
- Change the default value CMake parameter `JSON_MultipleHeader` to `ON` to always use the multi-header version unless specified otherwise. #3532
- Add option to disable default enum conversions (see [`JSON_DISABLE_ENUM_SERIALIZATION`](https://json.nlohmann.me/api/macros/json_disable_enum_serialization/)). #3536
- Add maintainer targets to create source archive `json.tar.xz` which can be used in [`FetchContent`](https://json.nlohmann.me/integration/cmake/#fetchcontent). #3289, #3255
- Fix `CITATION.cff`. #3320 
- Use [The Pitchfork Layout (PFL)](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs). #3462
- Update Cpplint. #3454
- More intermediate folders from Visual Studio Code are added to `.gitignore`. #3572
- Add badge for https://repology.org/project/nlohmann-json/versions. #3586
- Use [REUSE licensing framework](http://reuse.software). #3546 , #3633, #3635
- Install `.pc` and `.cmake` files to `share` directory. #3619

#### CI

- Add support for Visual Studio 2022., #3295
- Adjust GitHub Actions CI wrt. `windows-latest` image. #3368
- Remove deprecated `windows-2016` image. #3416
- Speed up AppVeyor CI. #3422 
- Update CI image to check with ICPC, GCC 11 and Clang 14. #3420
- Add build step for ICPC to CI. #3465
- Add more Xcode versions (13.3.1, 13.3, 13.2.1, 13.2, and 13.1) to the CI. #3485
- Fix "JSON_MultipleHeaders" option spelling in CI. #3555
- Increase timeout for Unicode tests. #3579, #3580, #3614
- Abort CI runs on newer commits on the same branch. #3610
- Fix MinGW CI failures. #3618
- Disable exceptions on ICPC. #3621

#### Unit tests

- Improve unit tests. #3380, #3393, #3365, #3405, #3377, #3421, #3422
- Disable regression test on GCC <8.4. #3451
- Add tests for 32 bit. #3524, #3529, #3530, #3532
- Add error message if test suite cannot be found. #3584, #3585
- Add unit test to make sure `iterator_input_adapter` advances iterators correctly. #3548
- Add a unit test including `<windows.h>`. #3631

#### Documentation and examples

- Add more examples to documentation. #3464
- Improve documentation on supported types of [`parse`](https://json.nlohmann.me/api/basic_json/parse/) and [`accept`](https://json.nlohmann.me/api/basic_json/accept/). #3245, #3246
- Add documentation on [how to parse JSON Lines](https://json.nlohmann.me/features/parsing/json_lines/) input. #3247
- Fix typos. #3249, #3265, #3426, #3439, #3481, #3499
- Document [parsing to `ordered_json`](https://json.nlohmann.me/features/object_order/#notes-on-parsing). #3325, #3326 
- Use `FetchContent_MakeAvailable` in examples. #3345, #3351
- Document [requirement](https://json.nlohmann.me/api/macros/json_diagnostics/#notes-on-parsing) of using the same definition of `JSON_DIAGNOSTICS` in all linked objects. #3360, #3378
- Document [fuzz testing](https://github.com/nlohmann/json/blob/develop/tests/fuzzing.md). #3477, #3478
- Add documentation of [macros](https://json.nlohmann.me/api/macros/) and [runtime assertions](https://json.nlohmann.me/features/assertions/). #3444, #3431
- Fix documentation of [JSON Pointer](https://json.nlohmann.me/features/json_pointer/). #3511, #3520 
- Make certain usage patterns more prominent in the README. #3557
- Document [precondition](https://json.nlohmann.me/api/basic_json/parse/#notes) for parsing from `FILE *` and add assertion. #3593 
- Improve documentation. #3592
- Add documentation for [comparing `json` and `ordered_json`](https://json.nlohmann.me/api/basic_json/operator_eq/#notes). #3443, #3599
- Adjust JSON Pointer examples and add CI step for the examples. #3622
- Overwork documentation and add more examples. #3553

### :fire: Deprecated functions

The implicit conversion from JSON Pointers to string ([`json_pointer::operator string_t`](https://json.nlohmann.me/api/json_pointer/operator_string_t)) has been deprecated. Use [`json_pointer::to_string`](https://json.nlohmann.me/api/json_pointer/to_string/) instead.

The following functions have been deprecated in earlier versions and will be removed in the next major version (i.e., 4.0.0):

- Function [`iterator_wrapper`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a0a8051760196ac813fd5eb3c8d5a2976.html#a0a8051760196ac813fd5eb3c8d5a2976) are deprecated. Please use the member function [`items()`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a5961446010dfc494e0c247b4e9026977.html#a5961446010dfc494e0c247b4e9026977) instead.
- Functions [`friend std::istream& operator<<(basic_json&, std::istream&)`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a60ca396028b8d9714c6e10efbf475af6.html#a60ca396028b8d9714c6e10efbf475af6) and [`friend std::ostream& operator>>(const basic_json&, std::ostream&)`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a34d6a60dd99e9f33b8273a1c8db5669b.html#a34d6a60dd99e9f33b8273a1c8db5669b) are deprecated. Please use [`friend std::istream&  operator>>(std::istream&, basic_json&)`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_aaf363408931d76472ded14017e59c9e8.html#aaf363408931d76472ded14017e59c9e8) and [`friend operator<<(std::ostream&, const basic_json&)`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a5e34c5435e557d0bf666bd7311211405.html#a5e34c5435e557d0bf666bd7311211405) instead.
- Passing iterator pairs or pointer/length pairs to parsing functions (`basic_json::parse`, `basic_json::accept`, `basic_json::sax_parse`, `basic_json::from_cbor`, `basic_json::from_msgpack`, `basic_json::from_ubjson`, `basic_json::from_bson`) via initializer lists is deprecated. Instead, pass two iterators; for instance, call `basic_json::from_cbor(ptr, ptr+len)` instead of `basic_json::from_cbor({ptr, len})`.

All deprecations are annotated with [`HEDLEY_DEPRECATED_FOR`](https://nemequ.github.io/hedley/api-reference.html#HEDLEY_DEPRECATED_FOR) to report which function to use instead.

**Comment by falbrechtskirchinger** [comment]
Can you please put this up as a Gist? I've quite a few notes.

Or create a PR.

**Reply by nlohmann** [reply]
Sure: https://gist.github.com/nlohmann/a6ac8cde82bde775a2baad29e758de25

**Reply by falbrechtskirchinger** [reply]
I'll be posting my changes here: https://github.com/falbrechtskirchinger/json_release_notes/pull/1

It will take a while, but I'll ping you once I'm done.

**Reply by falbrechtskirchinger** [reply]
FYI, you have a bunch of Doxygen links in there that don't work. Some can be replaced, others have no equivalent in the new documentation.

**Reply by nlohmann** [reply]
I’ll check later. Thanks for reviewing!

**Reply by falbrechtskirchinger** [reply]
The first pass is done. I've removed the dead Doxygen links and substituted mkdocs ones where possible.

**Reply by nlohmann** [reply]
Thanks everybody! The 3.11.0 is finally released!!!

**Reply by nlohmann** [reply]
FYI: https://github.com/nlohmann/json/discussions/3641
